src/gflownet/envs/seq_building_env.py

```diff
@@ -0,0 +1,133 @@
+from copy import deepcopy
+from typing import Any, List, Sequence, Tuple
+
+import torch
+from torch.nn.utils.rnn import pad_sequence
+from torch_geometric.data import Data
+
+from gflownet.envs.graph_building_env import (
+    Graph,
+    GraphAction,
+    GraphActionType,
+    GraphBuildingEnv,
+    GraphBuildingEnvContext,
+)
+
+
+# For typing's sake, we'll pretend that a sequence is a graph.
+class Seq(Graph):
+    def __init__(self):
+        self.seq: list[Any] = []
+
+    def __repr__(self):
+        return "".join(map(str, self.seq))
+
+    @property
+    def nodes(self):
+        return self.seq
+
+
+class SeqBuildingEnv(GraphBuildingEnv):
+    """This class masquerades as a GraphBuildingEnv, but actually generates sequences of tokens."""
+
+    def __init__(self, variant):
+        super().__init__()
+
+    def new(self):
+        return Seq()
+
+    def step(self, g: Graph, a: GraphAction):
+        s: Seq = deepcopy(g)  # type: ignore
+        if a.action == GraphActionType.AddNode:
+            s.seq.append(a.value)
+        return s
+
+    def count_backward_transitions(self, g: Graph, check_idempotent: bool = False):
+        return 1
+
+    def parents(self, g: Graph):
+        s: Seq = deepcopy(g)  # type: ignore
+        if not len(s.seq):
+            return []
+        v = s.seq.pop()
+        return [(GraphAction(GraphActionType.AddNode, value=v), s)]
+
+    def reverse(self, g: Graph, ga: GraphAction):
+        # TODO: if we implement non-LR variants we'll need to do something here
+        return GraphAction(GraphActionType.Stop)
+
+
+class SeqBatch:
+    def __init__(self, seqs: List[torch.Tensor], pad: int):
+        self.seqs = seqs
+        self.x = pad_sequence(seqs, batch_first=False, padding_value=pad)
+        self.mask = self.x.eq(pad).T
+        self.lens = torch.tensor([len(i) for i in seqs], dtype=torch.long)
+        # This tells where (in the flattened array of outputs) the non-masked outputs are.
+        # E.g. if the batch is [["ABC", "VWXYZ"]], logit_idx would be [0, 1, 2, 5, 6, 7, 8, 9]
```

Comment:
[minor] do you have an explicit <eos> token in your vocab ? Since the sequences are flattened and concatenated into a larger sequence, it might be useful to explicitely learn what define the end of a sequence (or something similar for the begining of sequence, well, any sequence separator really)